Vectara

ofermend opened the pull request:
# Vectara Integration

This PR provides integration with Vectara. Implemented here are:
* langchain/vectorstore/vectara.py
* tests/integration_tests/vectorstores/test_vectara.py
* langchain/retrievers/vectara_retriever.py
And two IPYNB notebooks to do more testing:
* docs/modules/chains/index_examples/vectara_text_generation.ipynb
* docs/modules/indexes/vectorstores/examples/vectara.ipynb

ofermend committed:
Initial commit

ofermend committed:
updates

ofermend committed:
Merge branch 'master' of https://github.com/ofermend/langchain

ofermend committed:
Merge branch 'master' of https://github.com/ofermend/langchain

ofermend committed:
added semi-upsert

ofermend committed:
updated docstrings

ofermend committed:
bug fixes; new notebook: chat_vectara

dev2049 commented on langchain/vectorstores/vectara.py:
```diff
@@ -0,0 +1,246 @@
+from langchain.embeddings.base import Embeddings
+from langchain.vectorstores.base import VectorStore, VectorStoreRetriever
+
+"""Implementation of Vector Store using Vectara (https://vectara.com)"""
```
docstring a bit out of place (if here should be comment)

ofermend replied:
> the two notebooks in chains/index_examples are a bit out place - we can't really have that for all integrations. if you want to keep around, i would move into the integrations directory and link to from the integrations page

Moved the notebooks to integrations. Thanks!

ofermend replied:
> docstring a bit out of place (if here should be comment)

Yep. fixed!

dev2049 commented on langchain/retrievers/vectara_retriever.py:
```diff
@@ -0,0 +1,36 @@
+from langchain.schema import BaseRetriever, Document
+from langchain.vectorstores.vectara import Vectara
+
+class VectaraRetriever(BaseRetriever):
```
we probably want to make Vectara.as_retriever() return this - so i would move it into that file (similar to how we do with redis)

ofermend replied:
so just move VectaraRetriever to vectara.py and get rid of vectara_retriever.py altogether?

dev2049 replied:
yep!

ofermend replied:
Perfect. Done!

dev2049 reviewed (commented):
the two notebooks in chains/index_examples are a bit out place - we can't really have that for all integrations. if you want to keep around, i would move into the integrations directory and link to from the integrations page

ofermend committed:
Updated docs (vectara.md)
Moved vectara_retriever code inside vectara.py per suggestion
Moved IPYNB files to docs/integration (and changed name of one of them)
Updated formatting to be compatible with "black"

ofermend committed:
Merge branch 'hwchase17:master' into vectara

ofermend committed:
updated vectara.md

ofermend committed:
Merge branch 'vectara' of https://github.com/ofermend/langchain into vectara

dev2049 committed:
fmt

ofermend committed:
remove unneded imports

ofermend committed:
remove unneded imports

dev2049 committed:
cr

dev2049 commented on langchain/vectorstores/vectara.py:
```diff
@@ -0,0 +1,328 @@
+        """Initialize with Vectara API."""
+        self._vectara_customer_id = vectara_customer_id
+        self._vectara_corpus_id = vectara_corpus_id
+        self._vectara_api_key = vectara_api_key
```
we usually make it possible to set credentials (like api key and other id's) via environment variables. if we wanted to support that here could d something like
```python
def __init__(vectara_api_key: Optional[str]=None, ...):
  self.vectara_api_key = vectara_api_key or get_from_env("vectara_api_key", "VECTARA_API_KEY")
  ...
```

ofermend replied:
Good suggestion - updated!

dev2049 commented on tests/integration_tests/vectorstores/test_vectara.py:
```diff
@@ -0,0 +1,28 @@
+    new_texts = ["foobar", "foobaz"]
+    docsearch.add_documents([Document(page_content=content) for content in new_texts])
+    output = docsearch.similarity_search("foobar", k=1)
+    assert output == [Document(page_content="foobar")] or output == [
```
is this not deterministic?

ofermend replied:
Good point. I updated the test to not only be deterministic but also more comprehensive and test metadatas.

dev2049 reviewed (commented):
two quick comments

dev2049 reviewed (commented):
two quick comments

dev2049 committed:
fmt

dev2049 committed:
import

ofermend committed:
updates to documentation
updated Vectara class to get api_keys, corpus_id and customer_id optionally from env
updated example notebooks

dev2049 committed:
fmt

dev2049 committed:
fix

ofermend committed:
minor updates for lint

ofermend committed:
Merge branch 'vectara' of https://github.com/ofermend/langchain into vectara

ofermend committed:
fixed lint issues

dev2049 committed:
fmt

dev2049 committed:
merge

dev2049 committed:
cr
